### ISSUE TITLE
Dictionaries insider lists in YAML variable files not converted to DotDict objects

### ISSUE THREAD

**Issue by jiangliangnihao2022** [robotframework/robotframework#4418]
Yaml file: 
```yaml
x:
  a: a0
  b: b0
y:
  - a: a1
    b: b1
  - a: a2
    b: b2
```

Robot file:
```robotframework
*** Setting ***
Variables      ./data.yaml

*** Test Cases ***
Import Variables By Yaml Test
    ${type_x}=    Evaluate    type($x)
    Log To Console    \nx type is ${type_x}

    ${y_0}=    Set Variable    ${y}[0]
    ${type_y_0}=    Evaluate    type($y_0)
    Log To Console    y[0] type is ${type_y_0}\n
```
Print log:                                        .
**x type is <class 'robot.utils.dotdict.DotDict'>
...y[0] type is <class 'dict'>****

 I think y[0] also should be DotDict.  would you fix this issue, because Variable y[0] can not use more feature like .(dot)

===advice===
update YamlImporter._set function in src/robot/variables/filesetter.py
before:
```python
   def _dot_dict(self, value):
       if is_dict_like(value):
           value = DotDict((n, self._dot_dict(v)) for n, v in value.items())
       return value
```
after:
```python
    def _dot_dict(self, value):
        if is_dict_like(value):
            value = DotDict((n, self._dot_dict(v)) for n, v in value.items())
        if isinstance(value, list):
            value = [ self._dot_dict(v) for v in value ]
        return value
```

**Comment by pekkaklarck** [comment]
True, the code doesn't currently take into account dicts inside lists. The proposed solution looks good but could possibly be changed to
```python
        if is_dict_like(value):
            return DotDict((n, self._dot_dict(v)) for n, v in value.items())
        if is_list_like(value):
            return [self._dot_dict(v) for v in value]
        return value
```
Using `is_list_like` instead of `isinstance` is consistent with `is_dict_like`. Alternatively both could use ´isinstance`, but then there could be problems is PyYAML would return some custom list/dict objects.

Are you @jiangliangnihao2022 interested to provide a PR? It would need tests in addition to the fix. Since this is a bug, I can fix it before RF 5.1 myself as well.

**Comment by jiangliangnihao2022** [comment]
Hi  Pekka,
I am glad to receive your quick reply. Thanks!
I also think your revision is more reasonable. So I'm sure you can do better.
About PR, I am not familiar with what is PR.


> True, the code doesn't currently take into account dicts inside lists. The proposed solution looks good but could possibly be changed to
> 
> ```python
>         if is_dict_like(value):
>             return DotDict((n, self._dot_dict(v)) for n, v in value.items())
>         if is_list_like(value):
>             return [self._dot_dict(v) for v in value]
>         return value
> ```
> 
> Using `is_list_like` instead of `isinstance` is consistent with `is_dict_like`. Alternatively both could use `isinstance`, but then there could be problems is PyYAML would return some custom list/dict objects.
> 
> Are you @jiangliangnihao2022 interested to provide a PR? It would need tests in addition to the fix. Since this is a bug, I can fix it before RF 5.1 myself as well.

Hi  Pekka,
I am glad to receive your quick reply. Thanks!
I also think your revision is more reasonable. So I'm sure you can do better.
About PR, I am not familiar with what is PR?  Is there any chance you can fix it directly?

**Comment by jiangliangnihao2022** [comment]
> True, the code doesn't currently take into account dicts inside lists. The proposed solution looks good but could possibly be changed to
> 
> ```python
>         if is_dict_like(value):
>             return DotDict((n, self._dot_dict(v)) for n, v in value.items())
>         if is_list_like(value):
>             return [self._dot_dict(v) for v in value]
>         return value
> ```
> 
> Using `is_list_like` instead of `isinstance` is consistent with `is_dict_like`. Alternatively both could use `isinstance`, but then there could be problems is PyYAML would return some custom list/dict objects.
> 
> Are you @jiangliangnihao2022 interested to provide a PR? It would need tests in addition to the fix. Since this is a bug, I can fix it before RF 5.1 myself as well.


Hi  Pekka,
I am glad to receive your quick reply. Thanks!
I also think your revision is more reasonable. So I'm sure you can do better.
About PR, I am not familiar with what is PR?  Is there any chance you can fix it directly?

**Comment by pekkaklarck** [comment]
PR means a pull request. They are the main way to contribute code to projects on GitHub. You can find more info about that from our [Contribution guidelines](https://github.com/robotframework/robotframework/blob/master/CONTRIBUTING.rst) and from GitHub documentation. As I already wrote, I can fix this myself as well.
